### ISSUE TITLE
Proposal: Option to include explanation (trace output) in decision logs

### ISSUE THREAD

**Issue by gshively11** [open-policy-agent/opa#2897]
I've posted previously about wanting to capture the reason why a policy succeeds or fails. I've tried to encode the reasons in my policy directly, but the policy I built became very verbose and hard to follow. It also more than doubled my query eval times.

I'm just now starting to play around with the built in `trace()` functionality and `explanation=notes`, and I'm seeing some potential in this feature to meet my needs.  However, it doesn't appear that the explanation field shows up in the decision logs, which is ultimately where I need the policy decision reasons, for auditing purposes.

Could we add another querystring flag, or a flag somewhere else, that will result in explanation showing up in the decision logs?  And also, do you foresee any issues using `trace()` to capture decision reasons?

**Comment by gshively11** [comment]
Actually, I think it would be better to have the flag exist on the decision log config, `include_explanation: true`.

**Comment by tsandall** [comment]
Some initial thoughts...

For most realistic policies, including the trace/explanation on every single decision log record is not feasible because of the size of the trace. Tracing also comes with substantial cost. The cost can be reduced by not including the values of variables in the expressions, but that lowers the values of the trace (e.g., understanding why the expression `input.foo == x` failed is not possible without knowing the value of `x`). Because of these facts, I'm not sure we want to include this kind of option (e.g., users will run into hard limits pretty quickly.)

As far as `trace()` goes...there are a few things to keep in mind:

* If tracing is not enabled, these statements are ignored.

* We don't guarantee that `trace()` statements will be executed. E.g., if another expression in the rule body is executed before `trace()` and it fails, then the evaluator may never reach the `trace()` expression. Similarly, if the rule indexer decides to skip a rule, `trace()` statements inside the rule will not be executed.

**Comment by gshively11** [comment]
> For most realistic policies, including the trace/explanation on every single decision log record is not feasible because of the size of the trace.

It is certainly a little verbose, and I was thinking about trying to write another tracer that had a terser output.  But in our case, we're willing to deal with the size of the trace output, because it allows us to understand well enough why a policy succeeded/failed.  During auditing, we must be able to answer why any/every authorization decision was allowed/denied.

> Tracing also comes with substantial cost.

Can you elaborate a bit on this?  From my experimentation, tracing seems to be significantly faster than trying to code the policy in such a way that we can capture the reasons for success/failure. 

> understanding why the expression input.foo == x failed is not possible without knowing the value of x

In our case, it's enough to understand that a specific expression failed, not necessarily why it did, because we can usually determine the "why" by looking at the input, which is also collected in the decision logs.

> users will run into hard limits pretty quickly

Can you elaborate on this one as well?  I saw less resource usage with tracing than I did with creating a more verbose policy that tried to capture success/failure and propagate through a bunch of nested rules.

> If tracing is not enabled, these statements are ignored.

We are wrapping the opa server and are requiring tracing (explanation) to be enabled on all API requests.

> We don't guarantee that trace() statements will be executed. E.g., if another expression in the rule body is executed before trace() and it fails, then the evaluator may never reach the trace() expression. Similarly, if the rule indexer decides to skip a rule, trace() statements inside the rule will not be executed.

This is also acceptable for us, because we only care about the rule path taken by the decision.  So far, every test we've run with tracing has successfully shown us the data we need to understand why a decision was rendered.

**Comment by gshively11** [comment]
I think it would be helpful if I could show you the policy and the explanation output we're getting, but it contains too much private information to share in a public forum like this.  But using tracing, we're getting output that basically tells us things like:  "the jwt has the claim x, y, and z", "it's less than X hours old", "it's signed", "the user of the jwt has Z groups", etc.  we sort of use the trace messages like unique identifiers for rules also, where a certain trace message points us to a particular chunk of policy, if we need to dig into more specifics.

**Comment by tsandall** [comment]
> Can you elaborate a bit on this?

Full tracing requires the evaluator provide values for variables inside of the trace event. Internally these values are stored in a data structure that avoids copies but when exiting the evaluator they need to be copied so that the caller can have access to them in a simpler format. This copy is quite expensive compared to normal evaluation.

> Can you elaborate on this one as well?

Here's an example. If you have a trivial policy like this where the caller queries for example/allow:

```rego
package example

default allow = false

allow {
  data.actions[_] == input.action
}
```

The evaluator will produce one REDO event for every element in the `data.actions` collection. If the collection contains hundreds or thousands (or more) values, the trace will become prohibitively large. If OPA is configured to use async/HTTP-based decision logging, it will have to buffer those traces in-memory until it can upload.

> I think it would be helpful if I could show you the policy and the explanation output we're getting, but it contains too much private information to share in a public forum like this. But using tracing, we're getting output that basically tells us things like: "the jwt has the claim x, y, and z", "it's less than X hours old", "it's signed", "the user of the jwt has Z groups", etc. we sort of use the trace messages like unique identifiers for rules also, where a certain trace message points us to a particular chunk of policy, if we need to dig into more specifics.

It sounds like the rule tracing feature we've discussed before could help here. With rule tracing enabled the decision log records would contain a collection of rules that were defined/evaluated successfully during the query. You can think of this like a lo-fi version of the Coverage feature in the playground. The name of every rule that exited at least once would be included in the decision log. If you have rules built up around concepts like JWT validity and presence of claims, the decision log would indicate whether those conditions were met.

This approach would perform much better (probably to the point where it could be on-by-default) and it would scale much better (i.e., it would be independent of how much data/iteration was present in the policy.) The big assumption is that there is enough information encoded into the names of the rules. I wonder if it would be enough in your case. A quick way to sanity check would be to look at the trace() calls you're making.

**Comment by gshively11** [comment]
Thanks for explaining that. It does sound like the rule tracing feature would be enough for us, given that I'm essentially just re-creating the name of the rule in the `trace("")` calls within each rule. Any idea on how much effort it would take to implement the rule tracer? 

For now, we probably have to roll with the fork allowing us to include the trace output in the decision logs, since we really need that data and it doesn't seem to perform poorly for us (yet?). I have a few other things on my list I need to try to implement first before I could try to tackle the rule tracer (custom authorization for rest client used in plugins, signing key rotation with AWS KMS, and manual plugin triggers instead of timers).

**Comment by anderseknert** [comment]
@gshively11 sounds like what you'd want here is something like a `log` function to simply have something printed / persisted to the (decision) logs without keeping all the (possibly expensive) tracing information that comes with `trace`? I've been pondering that myself - especially when debugging a running a server with `--watch` and just want to print something quickly to know if I'm really losing my mind or if there might be another explanation as to why the line  I'm _sure_ should be evaluated isn't. Something like that could probably be useful both for debugging as well as for adding arbitrary data to logged decisions. The rule tracing feature sounds interesting nonetheless though.

**Comment by gshively11** [comment]
@anderseknert yea, a log function like that would work too.  having something that automatically recorded the names of the rules evaluated and their success/failure would be preferable, just because it would involve less boilerplate, but I'd be happy with either one.

**Comment by anderseknert** [comment]
Agreed, but they would not need to be mutually exclusive :)

**Comment by gshively11** [comment]
Closed, discussion will continue in #2089.
